include/proteus/Frontend/Func.hpp

```diff
@@ -90,120 +87,91 @@ class FuncBase {
-  Var &declVarInternal(StringRef Name, Type *Ty,
-                       Type *PointerElemType = nullptr);
-
-  template <typename T> Var &declVar(StringRef Name = "var") {
+  template <typename T> Var<T> declVarInternal(StringRef Name = "var") {
     static_assert(!std::is_array_v<T>, "Expected non-array type");
 
     Function *F = getFunction();
-    return declVarInternal(Name, TypeMap<T>::get(F->getContext()));
+    auto &Ctx = F->getContext();
+    Type *AllocaTy = TypeMap<T>::get(Ctx);
+    auto *Alloca = emitAlloca(AllocaTy, Name);
+
+    if constexpr (std::is_pointer_v<T>) {
+      Type *PtrElemTy = TypeMap<T>::getPointerElemType(Ctx);
+      return Var<T>(std::make_unique<PointerStorage>(Alloca, IRB, PtrElemTy),
+                    *this);
+    } else {
+      return Var<T>(std::make_unique<ScalarStorage>(Alloca, IRB), *this);
+    }
+  }
+
+  template <typename T> Var<T> declVar(StringRef Name = "var") {
+    static_assert(!std::is_array_v<T>, "Expected non-array type");
+    return declVarInternal<T>(Name);
   }
 
   template <typename T>
-  Var &declVar(size_t NElem, AddressSpace AS = AddressSpace::DEFAULT,
-               StringRef Name = "array_var") {
+  Var<T> declVar(size_t NElem, AddressSpace AS = AddressSpace::DEFAULT,
+                 StringRef Name = "array_var") {
     static_assert(std::is_array_v<T>, "Expected array type");
 
     Function *F = getFunction();
     auto *BasePointer =
         emitArrayCreate(TypeMap<T>::get(F->getContext(), NElem), AS, Name);
 
     auto *ArrTy = cast<ArrayType>(TypeMap<T>::get(F->getContext(), NElem));
-    auto &Ref = *Variables.emplace_back(
-        std::make_unique<ArrayVar>(BasePointer, *this, ArrTy));
-    return Ref;
+    return Var<T>(std::make_unique<ArrayStorage>(BasePointer, IRB, ArrTy),
+                  *this);
   }
 
-  template <typename T> Var &defVar(T Val, StringRef Name = "var") {
-    Function *F = getFunction();
-    Var &VarRef = declVarInternal(Name, TypeMap<T>::get(F->getContext()));
-    VarRef = Val;
-    return VarRef;
+  template <typename T> Var<T> defVar(T Val, StringRef Name = "var") {
+    Var<T> Var = declVarInternal<T>(Name);
+    Var = Val;
+    return Var;
   }
 
-  Var &defVar(const Var &Val, StringRef Name = "var") {
-    if (&Val.Fn != this)
-      PROTEUS_FATAL_ERROR("Variables should belong to the same function");
-    Var &VarRef = declVarInternal(Name, Val.getValueType());
-    VarRef = Val;
-    return VarRef;
+  template <typename T, typename U>
+  Var<T> defVar(const Var<U> &Var, StringRef Name = "var") {
+    auto Res = declVarInternal<T>(Name);
+    Res = Var;
+    return Res;
   }
 
   template <typename T>
-  Var &defRuntimeConst(T Val, StringRef Name = "run.const.var") {
-    Function *F = getFunction();
-    Var &VarRef = declVarInternal(Name, TypeMap<T>::get(F->getContext()));
-    VarRef = Val;
-    return VarRef;
+  Var<T> defRuntimeConst(T Val, StringRef Name = "run.const.var") {
```

Comment:
Open issue for future todo: create a `ConstVar` or `ImmutableVar` instead of the loosely check `defRuntimeConst` derived class with deleted assignment methods for static checking.